installer/x86_64/install.sh

```diff
@@ -486,29 +519,53 @@ fi
 # Add a menu entry for the DEMO OS
 # Note: assume that apparmor is supported in the kernel
 demo_grub_entry="$demo_volume_revision_label"
+if [ "$environment" = "sonic" ]; then
+    running_sonic_revision=$(cat /etc/sonic/sonic_version.yml | grep build_version | cut -f2 -d" ")
+    old_grub_menuentry=$(cat /host/grub/grub.cfg | sed "/$running_sonic_revision/,/}/!d")
+    demo_dev=$(echo $old_grub_menuentry | sed -e "s/.*root\=\(.*\)rw.*/\1/")
+    grub_menuentry=$(cat /host/grub/grub.cfg | sed "/menuentry ONIE/,/}/!d")
+fi
+
 cat <<EOF >> $grub_cfg
 menuentry '$demo_grub_entry' {
-        search --no-floppy --label --set=root $demo_volume_revision_label
-        echo    'Loading $demo_volume_revision_label $demo_type kernel ...'
+        search --no-floppy --label --set=root $demo_volume_label
+        echo    'Loading $demo_volume_label $demo_type kernel ...'
         insmod gzio
         if [ x$grub_platform = xxen ]; then insmod xzio; insmod lzopio; fi
         insmod part_msdos
         insmod ext2
-        linux   /boot/vmlinuz-3.16.0-4-amd64 root=$demo_dev rw $GRUB_CMDLINE_LINUX  \
-                loop=$FILESYSTEM_SQUASHFS loopfstype=squashfs                       \
+        linux   /$sonic_dir/boot/vmlinuz-3.16.0-4-amd64 root=$demo_dev rw $GRUB_CMDLINE_LINUX  \
+                loop=$sonic_dir/$FILESYSTEM_SQUASHFS loopfstype=squashfs                       \
                 apparmor=1 security=apparmor $ONIE_PLATFORM_EXTRA_CMDLINE_LINUX
-        echo    'Loading $demo_volume_revision_label $demo_type initial ramdisk ...'
-        initrd  /boot/initrd.img-3.16.0-4-amd64
+        echo    'Loading $demo_volume_label $demo_type initial ramdisk ...'
+        initrd  /$sonic_dir/boot/initrd.img-3.16.0-4-amd64
 }
 EOF
 
-# Add menu entries for ONIE -- use the grub fragment provided by the
-# ONIE distribution.
-$onie_root_dir/grub.d/50_onie_grub >> $grub_cfg
+if [ "$environment" != "sonic" ]; then
+    # Add menu entries for ONIE -- use the grub fragment provided by the
+    # ONIE distribution.
+    $onie_root_dir/grub.d/50_onie_grub >> $grub_cfg
+    mkdir -p $onie_initrd_tmp/$demo_mnt/grub
+else
+cat <<EOF >> $grub_cfg
+$old_grub_menuentry
+$grub_menuentry
+EOF
+fi
 
-mkdir -p $onie_initrd_tmp/$demo_mnt/grub
 cp $grub_cfg $onie_initrd_tmp/$demo_mnt/grub/grub.cfg
 
+# Remove extra SONiC image if any
+if [ "$environment" = "sonic" ]; then
+    for f in /$demo_mnt/$SONIC_IMAGE_DIR_PREFIX* ; do
+        if [ -d $f ] && [ "$f" != "/$demo_mnt/$SONIC_IMAGE_DIR_PREFIX-$running_sonic_revision" ] && [ "$f" != "/$demo_mnt/$sonic_dir" ]; then
+            echo "Removing old SONiC installation $f"
+            rm -rf $f
+        fi
+    done
+fi
```

Comment:
I suggest to move this logic before we do the installation, it will save some space.